### ISSUE TITLE
vertx-auth-jwt user.principal() can't get the right information

### ISSUE THREAD

**Issue by angela-1** [eclipse-vertx/vertx-auth#435]
### Questions

I use auth-mongo and auth-jwt to authenticate my user account.
These days I updated my vertx to `4.0.0.CR1` and the code below can't get right result when I called `user.principal()`.

code:

```java
jwtAuthProvider.authenticate(new JsonObject().put("jwt", token)).onSuccess(user -> {
  JsonObject credencial = user.principal(); // this line get wrong result!
  System.out.println("user profile " + user.principal().encodePrettily());
  routingContext.json(new JsonObject());
}).onFailure(err -> {
  routingContext.fail(403);
});
```

CR1 result:
```json
 {
  "access_token" : "<BASE64-TOKEN>"
}

```

But the `4.0.0.Beta3` version and older versions works fine. It can get right information in token.

Beta3 result:
```json
{
  "_id" : "5f222cb4193968243eef4feb",
  "name" : "user",
  "username" : "username",
  "roles" : [ "USER" ],
  "iat" : 1604977065,
  "exp" : 1604980665,
  "iss" : "Vert.x",
  "sub" : "API"
}
```

I don't know why, because it works fine in `Beta3` but wrong in `CR1`，is there something I missed?
Any help will be appreciated!

### Version

4.0.0.CR1

### Context

When I called `user.principal()` after `authenticate()`, it doesn't get right result like `4.0.0.Beta3`.

CR1 result:
```json
 {
  "access_token" : "<BASE64-TOKEN>"
}

```

beta3 result:
```json
{
  "_id" : "5f222cb4193968243eef4feb",
  "name" : "user",
  "username" : "username",
  "roles" : [ "USER" ],
  "iat" : 1604977065,
  "exp" : 1604980665,
  "iss" : "Vert.x",
  "sub" : "API"
}
```

### Do you have a reproducer?

No.

### Steps to reproduce

1. ...
2. ...
3. ...
4. ...

### Extra

* Anything that can be relevant such as OS version, JVM version

OS: Centos 8
JVM: openjdk 11.0.9

**Comment by pmlopes** [comment]
@angela-1 in order to get the user object with the same internal structure across modules. decoded token attributes are stored on the `attributes` holder. `principal` properties are read only immutable data, while the decoded/mutable ones live on `attributes`.

The goal was to pave the way to allow the oauth2 module to be a more flexible and feature complete drop in replacement for jwt in the future.

As this is a breaking change, it has been fixed on https://github.com/vert-x3/vertx-auth/commit/091df83b60fc531ef862003b2486710933fcb005 to still expose the token decoded properties on `principal` so most applications will keep running, but it is advisable that in the future a small migration from looking into the right place to be performed by users of 4.x.

TLDR; this has been fixed as a retro compatible fix and users are advised to use `attributes()` for decoded/mutable properties.

**Comment by angela-1** [comment]
@pmlopes I have tried the `attributes()` and get the right decoded information. But it still have one problem.

I use auth-mongo and auth-jwt, the auth-jwt now can get decoded information from `attributes()`, but the auth-mongo 
still gets decoded information from `principal()` and `attributes()` gets nothing.

`MongoAuth` and `JWTAuth` are both extend from `AuthProvider`, but now they have different behaviours, 
`JWTAuth` use `attributes()` and `MongoAuth` use `principal()`.

Will they have the same behavious in the future release, based on your comment above, to use `attributes()`? Thank you!

**Comment by pmlopes** [comment]
@angela-1 I've clearly documented what principal and attributes mean. I also added 2 helper methods that were discussed in the past to be added but for some reason never got in:

```java
User user = ...

if (user.containsKey("someKey")) {
  String someValue = user.get("someKey");
}
```

The `containsKey` will return true if the key is present on `attributes` **or** `principal`. The `get` will return if the `attributes` contains the key, else return the value if `principal` contains the key, or `null`.

**Comment by pmlopes** [comment]
So to clarify, yes mongo user may contain values on `principal` because those were returned from the mongo query. However any computed values should be in the attributes. The distinction is only relevant for the validation and security of the object but not so relevant for the end user.

The new helper methods are relevant for the end user and not so for the internal usage.

**Comment by angela-1** [comment]
@pmlopes Thank you very much.
